Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,339 +1,177 @@
+  .ofNat w (x.val <<< n)
+
+/-- The right-shift of a bitvector by some amount. -/
+protected def shiftRight (x : BitVec w) (n : Nat) : BitVec w :=
+  ⟨x.val >>> n, by
+      simp only [Nat.shiftRight_eq_div_pow]
+      exact lt_of_le_of_lt (Nat.div_le_self' _ _) (x.isLt) ⟩
+
+/-- Signed less than for bitvectors. -/
+protected def slt (x y : BitVec (w + 1)) : Prop :=
+  ((x.val >>> w = 1) ∧ (y.val >>> w = 0)) ∨ ((x.val >>> w = y.val >>> w) ∧ x < y)
+
+/-- Signed less than or equal to for bitvectors. -/
+protected def sle (x y : BitVec (w + 1)) : Prop :=
+  ((x.val >>> w = 1) ∧ (y.val >>> w = 0)) ∨ ((x.val >>> w = y.val >>> w) ∧ x ≤ y)
+
+/-- Signed greater than for bitvectors. -/
+protected def sgt (x y : BitVec (w + 1)) : Prop := BitVec.slt y x
+
+/-- Signed greater than or equal to for bitvectors. -/
+protected def sge (x y : BitVec (w + 1)) : Prop := BitVec.sle y x
+
+instance : Complement (BitVec w) := ⟨BitVec.complement⟩
+instance : AndOp (BitVec w) := ⟨BitVec.and⟩
+instance : OrOp (BitVec w) := ⟨BitVec.or⟩
+instance : Xor (BitVec w) := ⟨BitVec.xor⟩
+instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
+instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩
+
+/-- Left rotation by some amount. -/
+def rotateLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  x <<< n ||| x >>> (w - n)
+
+/-- Right rotation by some amount. -/
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Concatenation of two bitvectors. -/
+protected def append (x : BitVec w) (y : BitVec v) : BitVec (w + v) :=
+  ⟨x.val <<< v ||| y.val, Nat.add_comm _ _ ▸ Nat.append_lt y.isLt x.isLt⟩
+
+instance : HAppend (BitVec w) (BitVec v) (BitVec (w + v)) := ⟨BitVec.append⟩
+
+/-- Extract the bitvector between indices i and j. -/
+@[simp]
+def extract (i j) (x : BitVec w) : BitVec (i - j + 1) :=
+  BitVec.ofNat _ (x.val >>> j)
+
+/-- Generate and concatenate `i` copies of a bitvector. -/
+def replicate : (i : Nat) → BitVec w → BitVec (w * i)
+  | 0,   _ => 0
+  | n + 1, x =>
+    have hEq : w + w * n = w * (n + 1) := by
+      rw [Nat.mul_add, Nat.add_comm, Nat.mul_one]
+    hEq ▸ (x ++ replicate n x)
+
+/-- Pad a bitvector by the zero bitvector of length `i`. -/
+def zeroExtend (i : Nat) (x : BitVec w) : BitVec (w + i) :=
+  have hEq : w+i = i+w := Nat.add_comm _ _
+  hEq ▸ ((0 : BitVec i) ++ x)
+
+/-- Extend a bitvector by repeating the most significant bit `i` times. -/
+def signExtend (i) (x : BitVec w) : BitVec (w + i) :=
+  have hEq : ((w - 1) - (w - 1) + 1) * i + w = w + i := by
+    rw [Nat.sub_self, Nat.zero_add, Nat.one_mul, Nat.add_comm]
+  hEq ▸ ((replicate i (extract (w - 1) (w - 1) x)) ++ x)
+
+/-- Keep the `v` least significant bits of a bitvector. -/
+def shrink (v) (x : BitVec w) : BitVec v :=
+  if hZero : 0 < v then
+    have hEq : v - 1 + 0 + 1 = v := by
+      rw [Nat.add_zero]
+      exact Nat.sub_add_cancel hZero
+    hEq ▸ x.extract (v - 1) 0
+  else 0
```

Comment:
With my suggested change to `extract` this becomes
```suggestion
  x.extract v 0
```
which probably indicates you don't need the definition at all.